### DISCUSSION TITLE
CRUD classes - how to require a "save" method call?

### DISCUSSION THREAD

**Discussion by kkmuffme** [vimeo/psalm#10064] [Q&A]
e.g. https://psalm.dev/r/0d7a047291

This won't do anything because $l->save() wasn't called.

How can I make psalm give an error if a specific method "add" is called, that it should require a call to "save"?
In cases like this it's very obvious, but it sometimes gets forgotten in some code paths or when refactoring.

**Comment by weirdan** [comment]
I don't see how it could be checked without changing the definition of `SomeLib` (directly or using stubs). As far as I understand, you want to make sure that if `add()` is called at some point, then `save()` is called eventually on that same object (e.g. not necessarily in the same scope). It's this 'eventually' part that makes it complicated.

Possibly you could introduce a template variable to carry the state and then make assertions based on that, e.g. https://psalm.dev/r/4f349032a2
